litebird_sim/simulations.py

```diff
@@ -322,7 +322,7 @@ def flush(self):
 
     def create_observations(
         self,
-        detectors: list,
+        detectors: List[str],
```

Comment:
Good to have the type specified. I was wondering if it would be better to call this parameter `detector_names` to make it more explicit (the same for the `Observation` class): now that there is a `Detector` class, people might get confused and think they're supposed to pass a list of `Detector` objects here.

Reply:
I think that the type hint makes it already clear enough.

Reply:
I agree it's clearer, but when I wrote my comment this morning I've just realized that type hints are not visible when reading code like this:

```python
import litebird_sim as lbs

detectors = some_complicated_function()
sim = lbs.Simulation()
# Wait, is "detectors" a list of strings or of some other type?
sim.create_observation(detectors=detectors, ...)
```

This looks clearer:

```python
import litebird_sim as lbs

detector_names = some_complicated_function()
sim = lbs.Simulation()
sim.create_observation(detector_names=detector_names, ...)
```

Moreover, I expect that other functions/modules/classes will expect a list of `Detector` objects as inputs (e.g., to generate noise). If we use the same keyword `detectors=` for these two cases, this will make the code less clear:

```python
# This expects a list of strings…
sim.create_observation(detectors=detectors, ...)
# …but this wants a list of "Detector" object
generate_noise(detectors=detectors, ...)
```

I think that our code should use the following conventions:

- The keyword `detector_names` indicates that strings are expected;
- The keyword `detectors` indicates that `Detector` objects are expected.

This would imply we need to fix my implementation of `Observation.__init__` as well, but I think it's worth it.

Reply:
This is probably something that is more efficient to discuss at the call tomorrow. Very briefly, I think we should avoid passing Detectors, both for clarity and performance.

About clarity, I prefer if we pass only the arguments that the function needs instead of a Detector. This way the user knows exactly what is needed and can experiment with the function more easily (i.e. without creating a Detector object, which may be overcomplicated for the purpose). In general, I think we should strive for using as few new classes as possible: If a class can be reduced to a standard container, we should avoid defining new classes. Every new definition raises the barrier to the new users. In particular, it seems to me that we could use a simple dictionary (or SimpleNamespace) instead of a Detector.

About performance, all the examples that come to my mind run faster if the the arguments are passed with one array for each Detector attribute rather than a list of Detectors (or dictionaries). At least this is the idea behind this pull request.

I think that noise generation fits well both these arguments. I'd prefer the pattern to be

```
sim.create_observation(detectors=detectors, ...)
for obs in sim.observations:
    obs.tod += generate_noise(obs.fknee, obs.white_noise, ...)
```

Reply:
> This is probably something that is more efficient to discuss at the call tomorrow.

Given that not many people are going to attend, I would prefer to continue the discussion here, so that others can chime in if they want.


> About performance, all the examples that come to my mind run faster if the the arguments are passed with one array

You're right, I discovered that creating 100,000  detectors using one list of Detectors takes 90 ms, while creating separate NumPy arrays only takes 50 ms. However, I do not expect that users are going to notice if the code spends 0,04 seconds more in running a function that's only called once in the lifetime of a program… (Moreover, LiteBIRD is going to have far less than 100,000 detectors!)

> I prefer if we pass only the arguments that the function needs instead of a Detector.

As I already said somewhere else, I do like this idea of yours that the framework should provide functions with the explicit list of parameters. In fact, my proposal idea is compatible with this code of yours:

```python
sim.create_observation(detectors=detectors, ...)
for obs in sim.observations:
    obs.tod += generate_noise(obs.fknee, obs.white_noise, ...)
```

but I mistakenly wrote my example above forgetting `sim.` in front of `generate_noise` (I fixed my comment):

```python
# This expects a list of strings…
sim.create_observation(detectors=detectors, ...)
# …but this wants a list of "Detector" object
sim.generate_noise(detectors=detectors, ...) # Add "sim."
```

The idea is that there is a function, `generate_noise`, that takes the arguments `fknee_hz`, `white_noise_ksqrthz`, etc., exactly like in your example, plus a `Simulation.generate_noise` method that wraps `generate_noise` and uses the IMO to retrieve these parameters coherently. In this way, we have the best of the two worlds. (I am following the same approach in PR #48 : `Simulation.generate_pointing_information` is a high-level wrapper around `Observation.generate_pointing_information` that knows how to access the IMO to retrieve the scanning strategy parameters.)

>  In particular, it seems to me that we could use a simple dictionary (or SimpleNamespace) instead of a Detector.

My initial implementation of `Detector` was a simple named tuple, but I changed it to a class because in this way we can provide implementations for `Detector.__repr__`, `Detector.__str__`, etc. I agree that this is not as awesome as in Julia, where you can define even a `plot` function over your custom types (see [here](https://ziotom78.github.io/Harlequin.jl/stable/genpointings/) for an example about scanning strategies, and [here](https://lspestrip.github.io/Stripeline.jl/latest/instrumentdb/#Introduction-1) for bandpasses), but this can still be helpful when debugging codes or doing exploratory work..

Moreover, named tuples are immutable, and this might be too limiting. The best option for this kind of classes would be to use [data classes](https://docs.python.org/3/library/dataclasses.html#module-dataclasses), but they are not supported in Python 3.6. Regarding dictionaries, I strongly dislke them, because it's too easy for people to abuse them with code like:

    detector["noise_{}".format(measurement_unit)] = ...

I want to know exactly what are the names of the fields in a compound object, and this kind of code prevents me from doing so by just reading at it.

There are surely disadvantages in using a `Detector` class (more code to write, more verboseness, more classes to document…), but let's list some advantages as well:

-   If one must create a vector `fknee_hz`, a vector `white_noise_hz`, a vector of quaternions, etc., there is the increased chance that one of these vectors will have `N + 1` elements instead of `N` (or `N - 1`, or… you get the idea), expecially if they are initialized at different places in the code. Another bug, far more difficult to catch, is that all the lists have the same number of elements, but in a different order! This kind of bugs is impossible if you use a list of `Detector` objects. **This is by far my strongest point in favour of the adoption of a `Detector` class.**

    I learned this lesson the hard way several years ago, when I wrote a C code that passed arrays of different lengths to the parameters `ttype`, `tform`, and `tunit` in CFITSIO [`fits_create_tbl`](https://heasarc.gsfc.nasa.gov/docs/software/fitsio/c/c_user/node43.html); I kept getting `SIGSEGV` crashes, until I found that I missed one element in one of the three parameters. Now I'm a strong advocate of arrays of data structures for this kind of situations, even if this means that the code will take 0,04 s more to complete.

-   For objects that require a complex initialization, using a dedicated class can lead to improved clarity and performance. For instance, an object might want to initialize derived quantities from the parameters passed to its constructor, if these derived quantities are going to be used several times during the lifetime of the object and/or they are expensive to compute. For instance, unlike Healpy, Healpix.jl uses a [Resolution](https://github.com/ziotom78/Healpix.jl/blob/ebc24731c0275207e454af898e946d28864feddf/src/resolution.jl#L24) object instead of a plain `NSIDE` parameter to specify the resolution for a map, because it caches a number of quantities derived from `NSIDE` that should otherwise be computed over and over again. The class' constructor is the most obvious place where to place this kind of initialization code.

-   If some people are going to extend an existing code by plugging in some more modules, there is the chance that the code already has the inputs they require. Example: you want to include noise generation in some code that performs beam convolution simulations. If you're loading detectors from the IMO, the noise parameters are already in place in the `Detector` objects, and you just have to call `sim.generate_noise(detectors=detectors)`. The opposite case means that you must figure out how to retrieve this information, and you must be able to **construct an array of values that has the same order as the other arrays (e.g., detector quaternions) used elsewhere in the code** (see the first point above in the list).

-   It's true that creating a `Detector` instance is verbose, because the constructor expects many fields. However, this might be mitigated by specifying `None` as the default value for all its fields but `name` (this is something I already wanted to do). This is going to be useful in unit tests, where you typically want to use only a few parameters for a `Detector` object, because you could just write:

    ```python
    def test_noise():
        ...
        det = Detector(name="a", fknee_hz=1.0, wn_level_ksqrthz=10.0) # Forget about other parameters
        sim.generate_noise(det)
        # Now check that the noise looks good
    ```

    And, again, if you're just going to directly call functions like `generate_noise`, you do not need to create a `Detector` class at all:

    ```python
    obs = Observation(...)
    arr = lbs.generate_noise(fknee_hz=1.0, wn_level_ksqrthz=10.0, nsamples=obs.nsamples)
    ```

Reply:
> > About performance, all the examples that come to my mind run faster if the the arguments are passed with one array
> 
> You're right, I discovered that creating 100,000 detectors using one list of Detectors takes 90 ms, while creating separate NumPy arrays only takes 50 ms. However, I do not expect that users are going to notice if the code spends 0,04 seconds more in running a function that's only called once in the lifetime of a program… (Moreover, LiteBIRD is going to have far less than 100,000 detectors!)

I didn't have in mind the detector creation, but rather the usage of the information stored in the detectors. 

> 
> > I prefer if we pass only the arguments that the function needs instead of a Detector.
> 
> As I already said somewhere else, I do like this idea of yours that the framework should provide functions with the explicit list of parameters. In fact, my proposal idea is compatible with this code of yours:
> 
> ```python
> sim.create_observation(detectors=detectors, ...)
> for obs in sim.observations:
>     obs.tod += generate_noise(obs.fknee, obs.white_noise, ...)
> ```
> 
> but I mistakenly wrote my example above forgetting `sim.` in front of `generate_noise` (I fixed my comment):
> 
> ```python
> # This expects a list of strings…
> sim.create_observation(detectors=detectors, ...)
> # …but this wants a list of "Detector" object
> sim.generate_noise(detectors=detectors, ...) # Add "sim."
> ```
> 
> The idea is that there is a function, `generate_noise`, that takes the arguments `fknee_hz`, `white_noise_ksqrthz`, etc., exactly like in your example, plus a `Simulation.generate_noise` method that wraps `generate_noise` and uses the IMO to retrieve these parameters coherently. In this way, we have the best of the two worlds. (I am following the same approach in PR #48 : `Simulation.generate_pointing_information` is a high-level wrapper around `Observation.generate_pointing_information` that knows how to access the IMO to retrieve the scanning strategy parameters.)

Not sure it is worth discussing, as it all starts from an example meant for something else. I've got the impression that `generate_noise` or any other similar function don't need detectors as argument, as the simulation and the observations in it already know the detectors they are handling
> 
> > In particular, it seems to me that we could use a simple dictionary (or SimpleNamespace) instead of a Detector.
> 
> My initial implementation of `Detector` was a simple named tuple, but I changed it to a class because in this way we can provide implementations for `Detector.__repr__`, `Detector.__str__`, etc. I agree that this is not as awesome as in Julia, where you can define even a `plot` function over your custom types (see [here](https://ziotom78.github.io/Harlequin.jl/stable/genpointings/) for an example about scanning strategies, and [here](https://lspestrip.github.io/Stripeline.jl/latest/instrumentdb/#Introduction-1) for bandpasses), but this can still be helpful when debugging codes or doing exploratory work..

What about subclassing a dictionary (or SimpleNamespace) and overriding repr and str?
I think that it is important for the users to minimize the number of classes we introduce. I am sure that they will find themselves more comfortable when working with data structures and types they are already familiar with.

I generally agree with the pros you list. About a couple of them, we'll need anyway (bug-free) tools to load and stack the information of the detectors into arrays of the right length and order (that's what this PR is about).

Reply:
> What about subclassing a dictionary (or SimpleNamespace) and overriding repr and str?
I think that it is important for the users to minimize the number of classes we introduce. I am sure that they will find themselves more comfortable when working with data structures and types they are already familiar with.

I am more convinced by the familiarity with data types than by the purpose of minimizing the number of classes we define. (This probably comes from my C++ background, where you're encouraged to create as many classes as the concepts you're dealing with; and Python classes are far easier to declare and use than C++ classes!) However, I bet that everybody going to use our framework is already familiar with the concept of `class`, so `Detector` should cause no harm! :-)

I would avoid dictionaries like the plague, as they can lead to unreadable code (see `detector["noise_{}".format(measurement_unit)]`). The `SimpleNamespace` type is more appealing, but it has a few drawbacks (a few of them shared with a `dict` type):

1. I said above that using many classes is the way C++ programmers write codes. However, this is a deceiving comment, as the way recent versions of Python have evolved seems to acknowledge the importance of quickly creating new classes. In fact, Python 3.7 has introduced [data classes](https://docs.python.org/3/library/dataclasses.html), which seems to have been warmly welcomed by the Python community (see several [blog](https://realpython.com/python-data-classes/) [posts](https://towardsdatascience.com/data-classes-in-python-8d1a09c1294b)) and [HN](https://news.ycombinator.com/item?id=16868296) [comments](https://news.ycombinator.com/item?id=17074783). (I expect that, once we drop support for Python 3.6 at the end of 2021, we should redefine types like `ScanningStrategy` and `Detector` as data classes.)

2. If we decide to stick with «familiar» data types, I fear that `SimpleNamespace` is not familiar at all to the average user of our framework. In fact, it is *never* used in the [TOAST](https://github.com/hpc4cmb/toast), [s4cmb](https://github.com/JulienPeloton/s4cmb), [healpy](https://github.com/healpy/healpy/tree/master/healpy), [PySM](https://github.com/bthorne93/PySM_public) repositories, which are probably quite representative of the kind of code a CMB scientist reads regularly. (The only example I found was *one* line of code in your own [FGBuster](https://github.com/fgbuster/fgbuster/blob/master/fgbuster/observation_helpers.py#L234).) If you google `python simplenamespace`, you get 31,000 results, but if you google `python "data class"`, you get 219,000,000 results!

3. IMO, it provides too much flexibility. The list of fields in a `detector` variable is supposed to be known in advance, as they are the fields our framework is reading from the IMO and is publishing to the user. A `SimpleNamespace` or a `dict` is meant to be used when you want to add/remove keys during runtime; this is not the case for `Detector`.

   In fact, the very [Python documentation](https://docs.python.org/3/library/types.html#types.SimpleNamespace) suggests that you should prefer a `namedtuple` (my first choice for `Detector`, as I said above) over `SimpleNamespace` when you want a structured type, as it is the case for `Detector`.

4. Probably because of its flexibility, a `dict` or a `SimpleNamespace` is hard to document. You can associate a docstring to a `Detector` class, explaining the meaning of each field, the assumptions, examples about how to use its fields, etc. Is the same possible with a `dict` or a `SimpleNamespace`?

5. A `dict` or a `SimpleNamespace` does not allow you to properly specify a type in the declaration of functions and methods: the meaning of the parameter `dets: List[Detector]` is immediate, and if you have doubts about what a `Detector` is, just type `?Detector` in the REPL. How should we declare the parameter `dets` if we use a `dict` or a `SimpleNamespace`? Should we write `dets: List[Dict[str, Any]]` and then add  the list of needed keys to the docstring of that function (of *every* function taking a list of detectors, with the problem of preserving consistency among the comments)?